wolfcrypt/src/hpke.c

```diff
@@ -0,0 +1,1139 @@
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1335, USA
+ */
+
+#ifdef HAVE_CONFIG_H
+    #include <config.h>
+#endif
+
+#include <wolfssl/wolfcrypt/settings.h>
+
+#if defined(HAVE_HPKE) && (defined(HAVE_ECC) || defined(HAVE_CURVE25519)) && \
+    defined(HAVE_AESGCM)
+
+#include <wolfssl/wolfcrypt/error-crypt.h>
+#include <wolfssl/wolfcrypt/ecc.h>
+#include <wolfssl/wolfcrypt/curve25519.h>
+#include <wolfssl/wolfcrypt/hmac.h>
+#include <wolfssl/wolfcrypt/hash.h>
+#include <wolfssl/wolfcrypt/sha256.h>
+#include <wolfssl/wolfcrypt/sha512.h>
+#include <wolfssl/wolfcrypt/aes.h>
+#include <wolfssl/wolfcrypt/hpke.h>
+
+const int hpkeSupportedKem[HPKE_SUPPORTED_KEM_LEN] =
+{
+  DHKEM_P256_HKDF_SHA256,
+  DHKEM_P384_HKDF_SHA384,
+  DHKEM_P521_HKDF_SHA512,
+  DHKEM_X25519_HKDF_SHA256,
+};
+
+const int hpkeSupportedKdf[HPKE_SUPPORTED_KDF_LEN] =
+{
+    HKDF_SHA256,
+    HKDF_SHA384,
+    HKDF_SHA512,
+};
+
+const int hpkeSupportedAead[HPKE_SUPPORTED_AEAD_LEN] =
+{
+    HPKE_AES_128_GCM,
+    HPKE_AES_256_GCM,
+};
+
+static const char* KEM_STR = "KEM";
+static const int KEM_STR_LEN = 3;
+
+static const char* HPKE_STR = "HPKE";
+static const int HPKE_STR_LEN = 4;
+
+static const char* HPKE_VERSION_STR = "HPKE-v1";
+static const int HPKE_VERSION_STR_LEN = 7;
+
+static const char* EAE_PRK_LABEL_STR = "eae_prk";
+static const int EAE_PRK_LABEL_STR_LEN = 7;
+
+static const char* SHARED_SECRET_LABEL_STR = "shared_secret";
+static const int SHARED_SECRET_LABEL_STR_LEN = 13;
+
+static const char* PSK_ID_HASH_LABEL_STR = "psk_id_hash";
+static const int PSK_ID_HASH_LABEL_STR_LEN = 11;
+
+static const char* INFO_HASH_LABEL_STR = "info_hash";
+static const int INFO_HASH_LABEL_STR_LEN = 9;
+
+static const char* SECRET_LABEL_STR = "secret";
+static const int SECRET_LABEL_STR_LEN = 6;
+
+static const char* KEY_LABEL_STR = "key";
+static const int KEY_LABEL_STR_LEN = 3;
+
+static const char* BASE_NONCE_LABEL_STR = "base_nonce";
+static const int BASE_NONCE_LABEL_STR_LEN = 10;
+
+static const char* EXP_LABEL_STR = "exp";
+static const int EXP_LABEL_STR_LEN = 3;
+
+/* encode n as a byte string with length w, return 0 or error */
+static int I2OSP(int n, int w, byte* out)
+{
+    int i;
+
+    if (w <= 0 || w > 32) {
+        return MP_VAL;
+    }
+
+    /* if width is less than int max check that n is less than w bytes max */
+    /* if width is greater than int max check that n is less than int max */
+    if ((w < 4 && n > ((1 << (w * 8)) - 1)) || (w >= 4 && n > 0x7fffffff)) {
+        return MP_VAL;
+    }
+
+    /* make sure the byte string is cleared */
+    XMEMSET( out, 0, w );
+
+    for (i = 0; i < w && n > 0; i++) {
+        out[w-(i + 1)] = (byte)n;
```

Comment:
not a big issue as of now, but curious if big endian systems have been taken into account with these transformations?